### ISSUE TITLE
Multiples kool.services files

### ISSUE THREAD

**Issue by rrcfesc** [kool-dev/kool#294]
**Describe the feature**

I want to use multiple kool-service files like docker-compose.yml, on my team multiples users are using diferents SO like MAC, Windows and Linux, and We have slightly differences, supported by docker 

**Backward compatibility concerns**

Would this be BC? **No**

**Use cases**

- network
- volumes

**Examples**

docker-compose.yml/kool.services.yml
```yaml
version: '3'
services:
  web:
    image: php:5.6-apache
    expose:
      - "9000"
      - "3315"
    ports:
      - "8085:80"
    environment:
      PHP_XDEBUG_ENABLED: 1
      XDEBUG_CONFIG: remote_host=172.18.19.1 #static IP
```

docker-compose-windows.yml

```
version: '3'
services:
  web:
    environment:
      XDEBUG_CONFIG: remote_host=host.docker.internal
```

On docker I can use:

`docker-compose -f docker-compose.yml -f docker-compose-windows.yml up -d` for windows users
`docker-compose up -d` Linux Users

**Comment by fabriciojs** [comment]
@rrcfesc I see! Indeed that seems like a fair use case we could facilitate.

The first idea that comes to mind is: maybe we could before executing each `docker-compose` command execution, check for platform-specific `docker-compose-PLAT.yml` file (some naming pattern must be followed, the example you gave seems simple and good enough), and in case it exists we then append the `-f docker-compose-PLAT.yml`.

Maybe there are some exception we should think of. What do you think @dbpolito @danielsuguimoto ?

**Comment by dbpolito** [comment]
Maybe we should bring more functionality to kool.yml, we only have scripts there, what if we have things like, project name, docker compose files that that will be used for these?

```yml
name: something #optional
files:
  - docker-compose.yml
  - docker-compose.{{ APP_ENV }}.yml
scripts:
...
```

That gives tons of flexibility to have it hard coded or using envs, etc

**Comment by fabriciojs** [comment]
I am a fan of the idea of expanding `kool.yml` language to bring in more features, but at the same time, I am very wary about when, why, and how we follow down that path.

IMO *tons of flexibility* is followed by the complexity which in this case I believe we don't need to put in `kool.yml`. We already have `docker-compose.yml` as a requirement, so as this feature is totally on top of that file itself, I feel comfortable having it with the automatic platform mapping.

**Comment by dbpolito** [comment]
It can be more than 2 files, by platform, by env, etc.

Not sure we can cover much cases in this approach... And automatic means hidden sometimes, hard to discover / understand.

I'm ok with both approaches.

**Comment by fabriciojs** [comment]
@dbpolito I see your point.

I am feeling that it is simple and righteous enough at the moment to implement like I described, based on Ricardo own usage/solution. We will make it transparent that, in case you need to customize based on PLATFORM (one of `linux` or `darwin` or `windows`), you can just have the extra `docker-compose.PLATFORM.yml` file and we will pass that in.

Do you agree this is a good thing to have? If so I would plan to implement it soonish as it is simple.

**Comment by indiegenus** [comment]
@dbpolito @fabriciojs - I'm leaning towards Daniel's idea of using the `kool.yml` file to configure the loading of multiple Docker Compose files, as it seems more flexible, gives the developer/team more explicit control, and it doesn't introduce a new file naming convention.

I see it as a new root-level `config` key, along the lines of the following (and where `docker-compose.yml` is the default if no custom config is defined). Over time, other keys could be added under `config` to customize the Kool setup.

```
config:
    docker:
        compose:
            - docker-compose.yml
            - docker-compose.windows.yml
            - docker-compose.local.yml
scripts:
    ...
```

**Comment by fabriciojs** [comment]
@indiegenus @dbpolito 

I guess I was not clear, let me try again.

We certainly want to add in a more explicit configuration interface for developers using Kool. But how we do that, is a big deal. Not something to be decided in a side-issue context.

Just like Daniel suggested one way, Steve suggested another, I could propose here a hundred other ways of interfacing the configuration to the user, taking into consideration several possible factors. In all of these options, we need to assert what we are actually doing to DX and what use cases we are really helping.

So repeating myself to be clear, I am a fan of expanding configuration settings, but HOW we do that, is going to need a deeper discussion so we avoid adding a lot of frequent breaking changes and risk building something *odd*. Does everyone agree with that?

Given the above, I see two focus here:

1. Do we think on something limited and specific to the issue at hand, like my proposal, and implement?
2. Do we dive in now to the scenario of expanding configuration interface for the user - in a separe project/issue?

One does not exclude the other.

Thoughts?

**Comment by fabriciojs** [comment]
Hello @rrcfesc !

While we were talking about the possibilities in extending configuration so you could add in extra docker-compose.yml files, like in your suggested solution, we only now noticed that your problem could be solved by a different approach, one already well used within `kool` proposed `docker-compose.yml` files - **environment variables**.

It is the idea that you could have your *single*  `docker-compose.yml` file written like this:

```yaml
version: '3'
services:
  web:
    image: php:5.6-apache
    expose:
      - "9000"
      - "3315"
    ports:
      - "8085:80"
    environment:
      PHP_XDEBUG_ENABLED: 1
      XDEBUG_CONFIG: ${XDEBUG_CONFIG:-remote_host=172.18.19.1} #static IP

```

This would make that, Docker Compose would look for an actual environment variable named `XDEBUG_CONFIG` to pass in to the container, and in case it does not exist it would fall back to `remote_host=172.18.19.1` as its default value.

So someone on Windows that needs a different value could simply add to their `.env` file:

```.env
XDEBUG_CONFIG="remote_host=host.docker.internal"
```

And work with `kool start` normally. Working with environment variables placed on a `.env` file in the root of the project is an easy way to keep `docker-compose.yml` flexible, with out-of-the-box good defaults, but still allowing for super easy override where needed.

Do you think this would be an acceptable solution to your use case? Do you have any other use cases where extending the whole `docker-compose.yml` is the best way to go?

**Comment by rrcfesc** [comment]
Well, is right I can workaround with that solution or after some tests I can use:

```sh
$ docker-compose -f kool.service.yml -f docker-compose-windows.yml up -d
$ kool run whatevercommand
```

But I fund another case with mac, in this example is with the volumes:

```yaml
#kool.service.yml
version: '3'
services:
  web:
    image: php:7.1
    volumes:
      - ./symfony:/var/www/html/symfony
```

```yaml
#docker-compose-mac.yml
version: '3'
services:
  web:
    volumes:
      - project-native-osx-sync:/var/www/html:nocopy
volumes:
  lms-native-osx-sync:
    external: true
```      
as you can see this configuration is based on the performance of the volumes on MAC with Magento, a well know issue.

https://magento.stackexchange.com/questions/230891/magento-2-really-bad-performance-in-docker
http://docker-sync.io/


Whatever be the case I still can use docker-compose to override the configuration to fix that problem ,but, for some reason I can't  still detect in some cases when I use this workaround, and use Kool to setup the environment, the Kool network is stuck and I can't raise the network. Again I can't reproduce the case. I don't know if I that can help you?

**Comment by fabriciojs** [comment]
@rrcfesc I see.

So first, I believe the environment variables are indeed a well-mannered way of extending/customizing specific bits of your `docker-compose.yml` setup.

At times when that is not possible *in the fashion I previously shared *, turns out that Docker Compose itself comes out of the box with another *environment variable* that should solve this for you: `COMPOSE_FILE` variable - https://docs.docker.com/compose/reference/envvars/#compose_file.

> COMPOSE_FILE
> Specify the path to a Compose file. If not provided, Compose looks for a file named docker-compose.yml in the current directory and then each parent directory in succession until a file by that name is found.
>
> This variable supports multiple Compose files separated by a path separator (on Linux and macOS the path separator is :, on Windows it is ;). For example: COMPOSE_FILE=docker-compose.yml:docker-compose.prod.yml. The path separator can also be customized using COMPOSE_PATH_SEPARATOR.

So basically you can do the same thing I suggested for the `XDEBUG_CONFIG` with this `COMPOSE_FILE`, where only developers on Windows or Mac should be advised to uncomment/change their .env to include the necessary extra `docker-compose-plat.yml` files.

---

PS: about the `kool` network - the KOOL_GLOBAL_NETWORK is **external** because it is meant to be used across different projects running in the same host connecting them. Being *external* means it is expected to exist before you try to use it. When you do `kool start` we handle this check if exists/if not create it kind of logic for you seamlessly. Since you are using `docker-compose up -d` and not `kool start`, you probably sometimes didn't get the network created properly beforehand. This is another reason to possibly stick with `COMPOSE_FILE` variable and then keep using `kool start`, since *kool* value proposition is exactly to attempt to take off your attention these minor configuration nuances.

---

PS2: about the performance of Docker volumes on Mac. I am on Mac it's been a couple of years now, and when I moved from Linux to it, I was very sad about the Docker volumes performance. It was impracticable to run for example `yarn install` or `yarn watch` from within a container due to the delay in the volumes. That being said, it is been sometime now, I am using 100% containers to execute it all (with Docker default `delegated` volumes). They made some great improvements in the last couple of years. I am sure that docker-sync or mutagen.io still should offer a surplus on performance, but depending on the use case it is nice to test without them from time to time, to see if the Docker internal improvements got you to a reasonable state where you could then simplify your stack by removing the extra external volume handler.

**Comment by fabriciojs** [comment]
hey @rrcfesc - just wondering, have you considered/adopted the suggestion above about using `COMPOSE_FILE` environment variable? Do you feel that suffice for supporting your use case?

Let us know whether we should close this issue, or maybe consider a different solution. Thank you!
